Лаб. 4 6411 Шехматов Виктор

Nakamo312 opened the pull request:
**ФИО:** Шехматов Виктор 
**Номер группы:** 6411  
**Номер лабораторной:** 4  
**Номер варианта:** 25 
**Краткое описание предметной области:** Поликлиника.
**Краткое описание добавленных фич:** Добавлена распределенная система генерации тестовых данных через RabbitMQ. Три проекта: DataGenerator (библиотека генерации), RabbitMqProducer (отправка сообщений), RabbitMqConsumer (обработка и сохранение в БД). Система гарантирует порядок создания: специализации → пациенты → врачи → записи. Поддерживает подтверждения создания и возврат в очередь при ошибках..

alxmcs commented on BikeRental/.idea/.idea.BikeRental/.idea/encodings.xml:
Служебные файлы ide добавить в гитигнор

alxmcs commented on BikeRental/BikeRental.Application/Services/BikeService.cs:
```diff
@@ -0,0 +1,91 @@
+{
+    public async Task<Bike> Create(Bike newDto, CancellationToken cancellationToken)
+    {
+        var bike = new Bike
+        {
+            SerialNumber = newDto.SerialNumber,
+            Type = newDto.Type,
+            Model = newDto.Model,
+            Color = newDto.Color,
+            CostPerPrice = newDto.CostPerPrice
+        };
```
Зачем было тащить в проект автомаппер, если ты все делаешь вручную?
Либо пропиши мапы для всех сущностей и их dto, либо избавься от автомаппера полностью и сделай отдельные экстеншен классы с преобразованием сущностей в dto и обратно

alxmcs commented on BikeRental/BikeRental.Application/Services/BikeService.cs:
```diff
@@ -0,0 +1,91 @@
+{
+    public async Task<Bike> Create(Bike newDto, CancellationToken cancellationToken)
+    {
+        var bike = new Bike
+        {
+            SerialNumber = newDto.SerialNumber,
+            Type = newDto.Type,
+            Model = newDto.Model,
+            Color = newDto.Color,
+            CostPerPrice = newDto.CostPerPrice
+        };
+
+        await bikeRepository.Add(bike);
+        return new Bike
+        {
+            SerialNumber = bike.SerialNumber,
+            Type = bike.Type,
+            Model = bike.Model,
+            Color = bike.Color,
+            CostPerPrice = bike.CostPerPrice
+        };
```
Это вообще какая-то дичь - создаешь экземпляр доменного класса, инсертишь его в репозиторий, а потом возвращаешь новый экземпляр доменного класса с пропертями ранее созданного экземпляра. 
Зачем? Это полный бред

Причем он чуть ли не во всех методах происходит

alxmcs commented on BikeRental/BikeRental.Contracts/Customer/CustomerCreateUpdateDto.cs:
Здесь и во всех dto - нет xml комментариев

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Controllers/AnalyticalController.cs:
```diff
@@ -0,0 +1,165 @@
+    /// Вывести информацию обо всех спортивных велосипедах.
+    /// </summary>
+    [HttpGet("sport-bikes")]
+    public async Task<IActionResult> GetAllSportBikes(CancellationToken cancellationToken)
```
Здесь и далее по коду -  IActionResult используется, когда метод контроллера возвращает экземпляры разных классов на один и тот же метод. Это явно не твой случай

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Controllers/AnalyticalController.cs:
```diff
@@ -0,0 +1,165 @@
+    {
+        var bikes = await bikeService.GetList(cancellationToken);
+        var sportBikes = bikes.Where(b => b.Type == BikeType.Sport).ToList();
+        return Ok(sportBikes);
```
Здесь и далее по коду - Методы контроллера должны принимать в качетве входных параметров dto и возращать также dto. У тебя возвращаются доменные сущности

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Controllers/BikeController.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace BikeRental.RestApi.Host.Controllers;
+
+public class BikeController(ICrudService<Bike, Bike, int> crudService, ILogger<BikeController> logger)
```
Меня радует, что хоть кто-то переиспользовал лекционный код. Но это нужно делать вменяемым образом. В базовом классе в именах дженерик типов указано максимально очевидно, какие сущности надо подставлять:
```cs
public abstract class CrudControllerBase<TDto, TCreateUpdateDto, TKey>
```
Ты вместо `TDto` и `TCreateUpdateDto` вхуячил доменную сущность. Используй головной мозг, пожалуйста.

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Controllers/CustomerController.cs:
Причем тут все в порядке

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Controllers/RentController.cs:
И тут тоже все норм. Чем `BikeController` такой особенный?

Matvey6M6 replied:
тем что dto абсолютно ничем не будет отличаться от доменной сущности. Просто было лень плодить юзлес классы

alxmcs replied:
Как жаль, придется поплодить

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Program.cs:
```diff
@@ -0,0 +1,46 @@
+builder.Services.AddScoped<IRepository<Customer, int>, CustomerRepositoryMock>();
+builder.Services.AddScoped<IRepository<Rent, int>, RentRepositoryMock>();
+
+builder.Services.AddScoped<ICrudService<Bike, Bike, int>, BikeService>();
```
`BikeService`, очевидно, должна оперировать dto сущностями, а не dao

alxmcs commented on BikeRental/BikeRental.Contracts/IRepository.cs:
`IRepository` - это интерфейс доменной службы, его объявлению не место на аппликейшен слое

alxmcs commented on BikeRental/BikeRental.RestApi.Host/Program.cs:
```diff
@@ -0,0 +1,46 @@
+var mapper = mapperConfig.CreateMapper();
+builder.Services.AddSingleton(mapper);
+
+builder.Services.AddScoped<IRepository<Bike, int>, BikeRepositoryMock>();
+builder.Services.AddScoped<IRepository<Customer, int>, CustomerRepositoryMock>();
+builder.Services.AddScoped<IRepository<Rent, int>, RentRepositoryMock>();
```
Не забудь в 3 лабе убрать моки в обратно в тестовый проект, где им самое место, и заменить имплементацию репозиториев на ef core-ные

alxmcs commented:
Зачтено

Nakamo312 committed:
initial commit

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
add comments for IModel.cs

Nakamo312 committed:
change field IsRepeated in Appointment data model

Nakamo312 committed:
added Repository class. Added DataBase context class

Nakamo312 committed:
create Application Layer: -Added MappingProfile;\n-Added DTOs;

Nakamo312 committed:
created Base Controller and Services

Nakamo312 committed:
modified DTOs

Nakamo312 committed:
Add API Controllers

Nakamo312 committed:
Add InMmempry Repositories

Nakamo312 committed:
Added API CRUD endpoints, added DTO representation schemes, addrd logging through middleware

Nakamo312 committed:
update README.md

Nakamo312 committed:
changes status code and response header in Create methods

Nakamo312 committed:
Added analytic services and controllers

Nakamo312 committed:
rm file

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
remove exceptions messages from endpoints

Nakamo312 committed:
change swager gen documentation for DTO schemes. Change content-type for all endpoints

Nakamo312 committed:
refactor: decouple API layer from domain models

- Remove Domain.Model dependency from base controller
- Use reflection for ID access in DTOs instead of Model inheritance
- Keep API layer dependent only on DTOs and application contracts

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
fix regex in PatientUpdateDto

Nakamo312 committed:
Improve error management: hide detailed validation errors

Nakamo312 committed:
revert last changes

Nakamo312 committed:
change response stats code in DELETE methods

Nakamo312 committed:
Aspire mood orchestrator; creating a conference for the database

Nakamo312 committed:
update README.MD

Nakamo312 committed:
comments corrected

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
registred DataSeed in DI container

Nakamo312 committed:
update phone number regex

Nakamo312 committed:
added phone number normalizatiom; change contracts for PUT methods

Nakamo312 committed:
changes namespaces

Nakamo312 committed:
changes namespaces

Nakamo312 committed:
modified README.md

Nakamo312 committed:
change Domain Models type field from uint to Guid

Nakamo312 committed:
refactor Program.cs

Nakamo312 committed:
Rename tablenames and columns

Nakamo312 committed:
s

Nakamo312 committed:
Add project ClinicDataGenerator, Clinic.RabbitMqProducer,Clinc.RabbitMqConsumer

Nakamo312 committed:
Rename appsettingsDevelopment.json to appsettings.Development.json

danlla commented on Clinic/Clinic.DataGenerator/Services/Interfaces/ICreationConfirmable.cs:
```diff
@@ -0,0 +1,19 @@
+    /// Confirms successful entity creation and performs necessary cleanup
+    /// </summary>
+    /// <param name="dataHash">Unique hash identifying the entity data</param>
+    void ConfirmCreation(string dataHash);
```
Нужно проставить `public`
Ниже и в остальных интерфейсах во всем солюшене аналогично

danlla commented on Clinic/Clinic.DataGenerator/Services/AppointmentGenService.cs:
```diff
@@ -0,0 +1,91 @@
+            .RuleFor(a => a.IsRepeated, f => f.Random.Bool(0.3f))
+            .RuleFor(a => a.PatientId, f =>
+            {
+                if (!_currentPatientIds.Any())
```
Если используется `List`, то лучше использовать свойство `Count` для того, чтобы понять пустая ли коллекция
Ниже аналогично, ну и во всех местах где вызывается `Any()` у `List`

danlla commented on Clinic/Clinic.DataGenerator/Services/AppointmentGenService.cs:
```diff
@@ -0,0 +1,91 @@
+public class AppointmentGenService : IDataGenService<AppointmentCreateDto>
+{
+    private readonly Faker<AppointmentCreateDto> _faker;
+    private readonly List<Guid> _generatedIds = new();
```
```csharp
private readonly List<Guid> _generatedIds = [];
```
Ниже аналогично, ну и везде где создается коллекция и её тип уже известен лучше использовать `[]`

danlla commented on Clinic/Clinic.DataGenerator/Services/EntityIdTracker.cs:
```diff
@@ -0,0 +1,118 @@
+    /// <returns>List of created GUIDs for the specified entity type</returns>
+    public List<Guid> GetIds(string entityType)
+    {
+        return _createdIds.TryGetValue(entityType, out var idList) ? new List<Guid>(idList) : new List<Guid>();
```
Здесь например будет вот так:
```csharp
return _createdIds.TryGetValue(entityType, out var idList) ? [.. idList] : [];
```

danlla commented on Clinic/Clinic.DataGenerator/Services/EntityIdTracker.cs:
```diff
@@ -0,0 +1,118 @@
+            lock (dict)
+            {
+                if (dict.ContainsKey(dataHash))
+                    dict.Remove(dataHash);
```
Студия подсказывает: `Do not guard 'Dictionary.Remove(key)' with 'Dictionary.ContainsKey(key)'. The former already checks whether the key exists, and will not throw if it does not.`
И в этом она права, метод `Remove` сам проверит

danlla commented on Clinic/Clinic.DataGenerator/Services/SpecializationGenService.cs:
```diff
@@ -0,0 +1,112 @@
+            WriteIndented = false
+        });
+        using var sha256 = SHA256.Create();
+        var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(json));
```
```csharp
var hash = SHA256.HashData(Encoding.UTF8.GetBytes(json));
```

Аналогично ниже

danlla commented on Clinic/Clinic.DataGenerator/Services/SpecializationGenService.cs:
```diff
@@ -0,0 +1,112 @@
+        {
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+            WriteIndented = false
+        });
```
`JsonSerializerOptions` лучше вынести в отдельное поле/свойство
Аналогично ниже

danlla commented on Clinic/Clinic.RabbitMq.Producer/Clinic.RabbitMq.Producer.csproj:
```diff
@@ -0,0 +1,20 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Clinic.Application\Clinic.Application.csproj" />
```
Можно убрать, эта зависимость придет вместе с `Clinic.DataGenerator`

danlla commented on Clinic/Clinic.RabbitMq.Consumer/Services/EntityResponseService.cs:
```diff
@@ -0,0 +1,86 @@
+﻿using Clinic.Application.Dtos.RabbitMq;
+using Clinic.RabbitMq.Consumer.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using RabbitMQ.Client;
+using System.Text.Json;
+using System.Text;
+
+/// <summary>
+/// Service for sending entity creation response messages
+/// </summary>
+public class EntityResponseService(
+        IConnectionFactory connectionFactory,
```
Не хватает namespace

danlla commented on Clinic/Clinic.RabbitMq.Consumer/Consumers/RabbitMqConsumer.cs:
```diff
@@ -0,0 +1,287 @@
+
+        if (_channel != null)
+        {
+            await _channel.CloseAsync();
```
Тут лучше прокинуть в метод `CloseAsync` `cancellationToken`

danlla reviewed (changes requested):
Нужно некоторые мелочи поправить

Nakamo312 committed:
Fix code review issues

Nakamo312 committed:
Merge branch 'lab4' of https://github.com/Nakamo312/enterprise-development into lab4

Nakamo312 committed:
update reference

danlla commented on Clinic/Clinic.DataGenerator/Services/AppointmentGenService.cs:
```diff
@@ -0,0 +1,91 @@
+            .RuleFor(a => a.IsRepeated, f => f.Random.Bool(0.3f))
+            .RuleFor(a => a.PatientId, f =>
+            {
+                if (!_currentPatientIds.Any())
```
Не поправлено с прошлого ревью, для `List` проперти `Count` производительнее чем метод `Any()`

Ниже аналогично, тут много мест таких (17)

danlla commented on Clinic/Clinic.RabbitMq.Consumer/Services/EntityResponseService.cs:
```diff
@@ -0,0 +1,88 @@
+    public static string ComputeDataHash(object obj)
+    {
+        var json = JsonSerializer.Serialize(obj);
+        using var sha256 = System.Security.Cryptography.SHA256.Create();
+        var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(json));
```
Не поправлено с прошлого ревью:

```csharp
var hash = System.Security.Cryptography.SHA256.HashData(Encoding.UTF8.GetBytes(json));
```

danlla commented on Clinic/Clinic.DataGenerator/Services/AppointmentGenService.cs:
```diff
@@ -0,0 +1,91 @@
+    /// Gets the list of generated IDs for tracking purposes
+    /// </summary>
+    /// <returns>List of generated GUIDs</returns>
+    public List<Guid> GetGeneratedIds() => _generatedIds.ToList();
```
```csharp
public List<Guid> GetGeneratedIds() =>[.. _generatedIds];
```

Аналогично в остальных сервисах

danlla commented on Clinic/Clinic.DataGenerator/Services/DataGeneratorService.cs:
```diff
@@ -0,0 +1,102 @@
+public class DataGeneratorService
+{
+    private readonly EntityIdTracker _idTracker;
+    private SpecializationGenService _specializationGenService;
+    private PatientGenService _patientGenService;
+    private DoctorGenService _doctorGenService;
+    private AppointmentGenService _appointmentGenService;
```
Нужно сделать `readonly`

danlla commented on Clinic/Clinic.DataGenerator/Services/EntityIdTracker.cs:
```diff
@@ -0,0 +1,116 @@
+    /// <param name="entity">The entity object</param>
+    public void RegisterExpectedEntity(string entityType, string dataHash, object entity)
+    {
+        var entities = _expectedEntities.GetOrAdd(entityType, _ => new Dictionary<string, object>());
```
```csharp
var entities = _expectedEntities.GetOrAdd(entityType, _ => []);
```

danlla reviewed (approved):
Эти замечания нужно будет поправить в 5 лабе
